### PR TITLE
[InstCombine] Fold `fmul X, -0.0` into `copysign(0.0, -X)`

### PR DESCRIPTION
**Pull request by dtcxzyw**
`fneg + copysign` is better than fmul for analysis/codegen.
godbolt: https://godbolt.org/z/eEs6dGd1G
Alive2: https://alive2.llvm.org/ce/z/K3M5BA

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/fmul.ll [hunk 1]
```diff
@@ -1268,13 +1268,95 @@ define half @mul_zero(half %x) {
...
+
+define <2 x float> @mul_neg_zero_nnan_ninf_vec(<2 x float> nofpclass(inf nan) %a) {
+; CHECK-LABEL: @mul_neg_zero_nnan_ninf_vec(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = fneg <2 x float> [[A:%.*]]
+; CHECK-NEXT:    [[RET:%.*]] = call <2 x float> @llvm.copysign.v2f32(<2 x float> <float 0.000000e+00, float poison>, <2 x float> [[TMP0]])
+; CHECK-NEXT:    ret <2 x float> [[RET]]
+;
+entry:
+  %ret = fmul <2 x float> %a, <float -0.0, float poison>
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fmul.ll:1360]: test mixed +0/-0/poison?

File: llvm/test/Transforms/InstCombine/fmul.ll [hunk 2]
```diff
@@ -1268,13 +1268,95 @@ define half @mul_zero(half %x) {
...
-; CHECK-NEXT:    [[R:%.*]] = fmul nnan half [[X:%.*]], 0xH8000
+; CHECK-NEXT:    [[TMP1:%.*]] = fneg nnan half [[X:%.*]]
+; CHECK-NEXT:    [[R:%.*]] = call nnan half @llvm.copysign.f16(half 0xH0000, half [[TMP1]])
 ; CHECK-NEXT:    ret half [[R]]
 ;
   %r = fmul nnan half %x, -0.0
   ret half %r
 }
+
+define float @mul_pos_zero_nnan_ninf(float nofpclass(inf nan) %a) {
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fmul.ll:1281]: The no-infs aren't relevant throughout?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 3]
```diff
@@ -814,8 +814,17 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
...
-  if (I.hasNoNaNs() && match(Op1, m_PosZeroFP())) {
+  // With no-nans:
+  // X * 0.0 --> copysign(0.0, X)
+  // X * -0.0 --> copysign(0.0, -X)
+  if (match(Op1, m_AnyZeroFP()) &&
+      isKnownNeverNaN(&I, /*Depth=*/0, SQ.getWithInstruction(&I))) {
+    if (match(Op1, m_NegZeroFP())) {
+      Op0 = Builder.CreateFNegFMF(Op0, &I);
+      // Canonicalize -0.0 to 0.0
+      Op1 = ConstantFoldUnaryOpOperand(Instruction::FNeg, cast<Constant>(Op1),
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:825]: I don't expect free constant fold calls to appear like this. Should visit copysign just simplify all constant signs it sees?

File: llvm/test/Transforms/InstCombine/fmul.ll (hunk 2 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fmul.ll:1281]: inf * 0 = nan: https://alive2.llvm.org/ce/z/cY29pY

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 3 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:825]: I will post a patch later.

File: llvm/test/Transforms/InstCombine/fmul.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fmul.ll:1281]: I always forget this one

File: llvm/test/Transforms/InstCombine/fmul.ll [hunk 4]
```diff
@@ -1268,13 +1268,95 @@ define half @mul_zero(half %x) {
...
+}
+
+define float @mul_pos_zero_nnan_ninf_fmf(float nofpclass(nan) %a) {
+; CHECK-LABEL: @mul_pos_zero_nnan_ninf_fmf(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[RET:%.*]] = fmul ninf float [[A:%.*]], 0.000000e+00
+; CHECK-NEXT:    ret float [[RET]]
+;
+entry:
+  %ret = fmul ninf float %a, 0.000000e+00
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fmul.ll:1310]: This case should have folded? I also don't see where the infinity check is happening

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fmul.ll:1310]: Yeah, it should be folded. We can improve `computeKnownFPClass` to fix this. But I don't think this pattern exists in real-world applications.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fmul.ll:1310]: Why is any improvement needed? You just need to check the ninf flag?

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fmul.ll:1310]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 5]
```diff
@@ -814,8 +814,19 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
...
+  // X * 0.0 --> copysign(0.0, X)
+  // X * -0.0 --> copysign(0.0, -X)
+  const APFloat *FPC;
+  if (match(Op1, m_APFloatAllowUndef(FPC)) && FPC->isZero() &&
+      ((I.hasNoInfs() &&
+        isKnownNeverNaN(Op0, /*Depth=*/0, SQ.getWithInstruction(&I))) ||
+       isKnownNeverNaN(&I, /*Depth=*/0, SQ.getWithInstruction(&I)))) {
+    if (FPC->isNegative())
+      Op0 = Builder.CreateFNegFMF(Op0, &I);
+    Op1 = Constant::replaceUndefsWith(
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:827]: Alive2: https://alive2.llvm.org/ce/z/GNc4gb

**Review by arsenm — APPROVED** [review]

**Pull request merged**